### PR TITLE
resource store: save stage of resource creation and report to callers

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
Often, users are met with the error "Kubelet may be retrying requests that are timing out in CRI-O..." but aren't provided with any details as to why.

This PR adds functionality that saves the stage of container and pod creations, as well as logs them. Now, users will be able to see where their pod or container is
#### Does this PR introduce a user-facing change?

```release-note
CRI-O now logs the stage of container or pod creation under system load. This allows users to find why their creation requests are stalling.
```

### CODE REVIEW

**Comment by haircommander** [comment]
/hold

I don't think a single map is the best choice here. we may need something shareded. need to do a perf analysis under load.
also need tests here


still, I wanted to spread to a wider audience so folks can chime in on stages that shoudl be added or removed.

**Comment by rphillips** [comment]
can we move the logic into the Resource?

File: internal/resourcestore/resourcestore.go
Comment (haircommander) [inline internal/resourcestore/resourcestore.go:27]: my current thouht is separating the locking mechanisms will reduce contention. still, if there are hundreds of containers or pods being created then this will get very busy. may need something sharded by key so each item can be updated atomically without bothering the others...

**Comment by haircommander** [comment]
> can we move the logic into the Resource?

I thought about it, see https://github.com/cri-o/cri-o/pull/5683#discussion_r814297858 on why I decided doing it this way. I'm open if we can find a performant way

File: internal/resourcestore/resourcestore.go
Comment (rphillips) [reply internal/resourcestore/resourcestore.go:27]: migrating to a RWLock might have some benefit.

Comment (haircommander) [reply internal/resourcestore/resourcestore.go:27]: unfortunately, this value is write often, and rarely read. I have thought about using a RW mutex on the whole map and only adding names when we have a write lock, and then having an [atomic value](https://pkg.go.dev/sync/atomic#Value) for the Stage...

**Comment by saschagrunert** [comment]
A friendly reminder that this PR had no activity for 30 days.

**Comment by haircommander** [comment]
@cri-o/cri-o-maintainers PTAL, I have revitalized this and reimplemented the `stages` map to handle concurrency better

**Comment by haircommander** [comment]
per reviews from both @rphillips and @kolyshkin I have just added the stage to Resource which simplifies things a bit

File: internal/resourcestore/resourcestore.go
Comment (kolyshkin) [inline internal/resourcestore/resourcestore.go:143]: This seem like a bugfix, so maybe put it as a separate commit.

File: internal/resourcestore/resourcestore.go
Comment (kolyshkin) [inline internal/resourcestore/resourcestore.go:189]: This does not belong to the first commit, as you still return a single value (i.e. the code can't be compiled).

File: server/utils.go
Comment (kolyshkin) [inline server/utils.go:199]: Should this belong to the second commit?

File: internal/resourcestore/resourcestore.go
Comment (kolyshkin) [inline internal/resourcestore/resourcestore.go:28]: This doesn't have to be global IMO.

File: internal/resourcestore/resourcestore.go
Comment (kolyshkin) [inline internal/resourcestore/resourcestore.go:220]: Seems that this method  is not used anywhere.

**Review by kolyshkin — CHANGES_REQUESTED** [review]

File: internal/resourcestore/resourcestore.go
Comment (haircommander) [reply internal/resourcestore/resourcestore.go]: I have left it for testing

**Comment by TomSweeneyRedHat** [comment]
LGTM assuming happy tests

**Comment by haircommander** [comment]
@saschagrunert PTAL

**Comment by haircommander** [comment]
wakeup prow

/retest

**Comment by haircommander** [comment]
prow was frozen, so I rebased, @saschagrunert can I have my tag back please :smiley:

**Comment by haircommander** [comment]
hm 
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/cri-o_cri-o/5683/pull-ci-cri-o-cri-o-main-ci-rhel-critest/1570759588955820032
cc @sohankunkerkar @wgahnagl 

/retest

File: internal/resourcestore/resourcestore.go
Comment (kolyshkin) [inline internal/resourcestore/resourcestore.go:195]: Perhaps you need to set the stage to unknown here.

Comment (kolyshkin) [reply internal/resourcestore/resourcestore.go:195]: ...and in every other place where you create a new `Resource`.

Alternatively, don't initialize it, but every time you return the stage value, check it for being empty and return `StageUnknown`. Is this is what you're doing, then nothing has to be changed in the code.

**Review by kolyshkin — APPROVED** [review]
LGTM

File: internal/resourcestore/resourcestore.go
Comment (mrunalp) [inline internal/resourcestore/resourcestore.go:40]: Can we make this a first class type or enum?

Comment (haircommander) [reply internal/resourcestore/resourcestore.go:40]: I had named strings a bit ago, but I think this is clearer and makes the source of the name closer to where it's relevant

**Pull request merged**
